### ISSUE TITLE
Runtime builder doesn't error when calling for more qubits than allocated

### ISSUE THREAD

**Issue by anthony-santana** [NVIDIA/cuda-quantum#684]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

Also see #683 

I'd expect the following code to throw an error during compilation as opposed to upon sample/observe call
```
import cudaq 

kernel = cudaq.make_kernel()
qubits = kernel.qalloc(5)

kernel.x(qubits[10])

print(kernel)
```
The reason I'd prefer a compilation error is because this produces invalid Quake
```
module {
  func.func @__nvqpp__mlirgen____nvqppBuilderKernel_270936062618() attributes {"cudaq-entrypoint"} {
    %0 = quake.alloca !quake.veq<5>
    %1 = quake.extract_ref %0[10] : (!quake.veq<5>) -> !quake.ref
    quake.x %1 : (!quake.ref) -> ()
    return
  }
}
``` 
which shouldn't be allowed.

### Steps to reproduce the bug

See description.

### Expected behavior

See description.

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: 
- **Python version**: 
- **C++ compiler**: 
- **Operating system**:

**Comment by bmhowe23** [comment]
I took a stab at resolving this in PR #688. Let me know what you think.

**Comment by bmhowe23** [comment]
Resolved by #688. The above example now produces the following:
```
>>> print(kernel)
loc("<builder>":1:1): error: 'quake.extract_ref' op invalid index [10] because >= size [5]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: cudaq::builder failed to JIT compile the Quake representation.
```
